### PR TITLE
Restart task

### PR DESCRIPTION
**Pull request by kirs**
This is an implementation of the restart task according to what we discussed in the original issue (https://github.com/Shopify/kubernetes-deploy/issues/43)

TL;DR there's no way to tell Kubernetes to restart the pods and that's something that we need to have control over processed in case of an incident.

The hackish way to get the pods restarted is to update the deployment with a noop ENV variable, like `RESTARTED_AT` containing a timestamp. This can be done with a kubectl command:

```
RESTART_DATE=$(date -u '+%Y-%m-%dT%H:%M:%S%z')
kubectl patch deployment web -n trashbin \
  -p '{"spec":{"template":{"spec":{"containers":[{"name":"app","env":[{"name":"RESTARTED_AT","value":"'$RESTART_DATE'"}]}]}}}}'
```

This PR implements the restart task as a CLI tool. The usage:

```
# restarts the "web" deployment
kubernetes-deploy-restart <kube namespace> <kube context> --deployments=web
```

Under the hood it crafts the `PATCH` request to update the deployment, similar to what the kubectl command does.

fixes https://github.com/Shopify/kubernetes-deploy/issues/43

review @sirupsen @KnVerey

### CODE REVIEW

File: exe/kubernetes-deploy-restart [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env ruby
```
Comment (sirupsen) [inline exe/kubernetes-deploy-restart]: What was the reason you decided to make a separate executable for this?

File: lib/kubernetes-deploy/restart_task.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      end
+    end
+
+    private
+
+    def build_v1beta1_kubeclient(context)
+      build_kubeclient(
+        api_version: "v1beta1",
+        context: context,
+        endpoint_path: "/apis/extensions/"
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: Aren't we always setting `/apis/extensions` when `api_version == 'v1beta1`? Can we do this automatically?

File: lib/kubernetes-deploy/restart_task.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        record.metadata.name,
+        build_patch_payload(record),
+        @namespace
+      )
+    end
+
+    def fetch_deployments(list)
+      list.map do |name|
+        records = @v1beta1_kubeclient.get_deployments(label_selector: "name=#{name}", namespace: @namespace)
+        raise "too many records: #{records.size}" if records.size > 1
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: This won't be unreasonably common at some point, right? Could we build some helpers to iterate through all the YAML segments?

File: lib/kubernetes-deploy/restart_task.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        records.first
+      end
+    end
+
+    def build_patch_payload(deployment)
+      containers = deployment.spec.template.spec.containers
+      if containers.size > 1
+        raise "don't know how to restart > 1 containers"
+      end
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: Same comment here as above, we should consider supporting this out the gate—it's going to take more time for someone to discover we don't need this, and add it in, than doing it now or as a follow-up I think 🤔

File: lib/kubernetes-deploy/restart_task.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+      {
+        spec: {
+          template: {
+            spec: {
+              containers: containers.map do |container|
+                {
+                  name: container.name,
+                  env: [{ name: "RESTARTED_AT", value: Time.now.to_i.to_s }]
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: Someone mentioned we can do this with annotations too, why did you choose environment variables?

File: lib/kubernetes-deploy/restart_task.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      if deployments.empty?
+        raise ArgumentError, "#perform takes at least one deployment to restart"
+      end
+
+      fetch_deployments(deployments).each do |record|
+        response = patch_deployment_with_restart(record)
+        if response.code == 200
+          @logger.info "Restarted #{record.metadata.name.inspect}"
+        else
+          @logger.fatal "Failed to restart #{record.metadata.name} (API returned non-200 response)"
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: Could you write the status code?

File: lib/kubernetes-deploy/restart_task.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    end
+
+    def perform(deployments)
+      if deployments.empty?
+        raise ArgumentError, "#perform takes at least one deployment to restart"
+      end
+
+      fetch_deployments(deployments).each do |record|
+        response = patch_deployment_with_restart(record)
+        if response.code == 200
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: I would've accepted a 202 to be honest. 🤔

File: test/integration/restart_test.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
+require 'test_helper'
+require 'kubernetes-deploy/restart_task'
+
+class KubernetesDeployTest < KubernetesDeploy::IntegrationTest
+  def test_restart
+    deploy_fixtures("basic", subset: ["configmap-data.yml", "web.yml.erb"])
```
Comment (sirupsen) [inline test/integration/restart_test.rb]: Why did you need the configmap?

File: test/integration/restart_test.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  private
+
+  def fetch_restarted_at(deployment)
+    containers = deployment.spec.template.spec.containers
+    assert_equal 1, containers.size
+    containers.first.env.find { |n| n.name == "RESTARTED_AT" }
+  end
+
+  def fetch_deployment(name)
```
Comment (sirupsen) [inline test/integration/restart_test.rb]: This seems like something we should write a or already have a helper for, no?

File: test/integration/restart_test.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
+require 'test_helper'
+require 'kubernetes-deploy/restart_task'
+
+class KubernetesDeployTest < KubernetesDeploy::IntegrationTest
+  def test_restart
```
Comment (sirupsen) [inline test/integration/restart_test.rb]: Could we get an error test case as well, or is that too hard to reproduce?

File: exe/kubernetes-deploy-restart (hunk 1 above)
Comment (kirs) [inline exe/kubernetes-deploy-restart]: Because it has completely different arguments and semantics, and it's more like a helper that we want to provide out of box.

File: lib/kubernetes-deploy/restart_task.rb (hunk 3 above)
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: That's actually an exceptional case (two deployments with name `web` in the same namespace). I'm not sure that can ever happen in production, but I'd prefer to fail rather than proceed.

File: lib/kubernetes-deploy/restart_task.rb (hunk 5 above)
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: I've looked into that a couple days ago and altering annotations or labels does not trigger the restart. Touching ENV looks like the only option. More details in https://github.com/Shopify/kubernetes-deploy/issues/43#issuecomment-284500937

File: test/integration/restart_test.rb (hunk 8 above)
Comment (kirs) [inline test/integration/restart_test.rb]: Because `web` from the fixtures is referring to config map values. That's how it works in all tests

File: lib/kubernetes-deploy/restart_task.rb (hunk 7 above)
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: I'll change that to 2**

File: lib/kubernetes-deploy/restart_task.rb (hunk 2 above)
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: No, because deployments are part of v1beta1 API and the rest is API v1.

Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: Ideally the API wrapper from the gem could do it automatically but it doesn't

File: test/integration/restart_test.rb (hunk 10 above)
Comment (kirs) [inline test/integration/restart_test.rb]: That's kind of hard to reproduce. I should have mentioned it as the reason why I didn't add an error case.

File: Gemfile [hunk 11]
```diff
@@ -4,5 +4,4 @@ source 'https://rubygems.org'
 gemspec
 
 gem 'pry'
-gem 'kubeclient'
```
Comment (kirs) [inline Gemfile]: This is now a gemspec dependency.

File: test/integration/restart_task_test.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+      namespace: @namespace,
+      logger: @logger
+    )
+    restart.perform(["web"])
+
+    assert_logs_match(/Restarted "web"/)
+
+    deployment = fetch_deployment("web")
+    assert fetch_restarted_at(deployment), "RESTARTED_AT is present after the restart"
+  end
```
Comment (sirupsen) [inline test/integration/restart_task_test.rb]: Is there any helper to test restarting multiple? What happens if I pass `web` twice?

File: lib/kubernetes-deploy/restart_task.rb [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+      @logger = logger
+      @v1beta1_kubeclient = build_v1beta1_kubeclient(context)
+    end
+
+    def perform(deployments)
+      if deployments.empty?
+        raise ArgumentError, "#perform takes at least one deployment to restart"
+      end
+
+      fetch_deployments(deployments).each do |record|
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: Does this mean each deployment is restarted in serial? How long does that take? Won't this only give the status that it's _starting_ to restart, not that it's actually successfully restarted all, say, 10,000 pods?

Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: You're right, I should have set the message to `triggering restart of "web"` because it doesn't way for the restart to finish. The pods are recreated according to the rollout strategy that every project has defined.

Do you think the tool needs the `--wait` flag to wait for the restart? I'm not sure about that because that's something that can be always watched in kubernetes dashboard.

File: test/integration/restart_task_test.rb (hunk 12 above)
Comment (kirs) [inline test/integration/restart_task_test.rb]: Good point. I'll add tests for that 👍

File: lib/kubernetes-deploy/restart_task.rb (hunk 13 above)
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: If it's hard to build, it's probably not worth it—but most devs would have no idea to look there and probably won't even have access. It would feel like you should be able to use the same code as we do for deploys to do this without too much trouble, but I'm not sure. I'll leave this up to you.

File: lib/kubernetes-deploy/restart_task.rb (hunk 2 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/restart_task.rb]: Related: https://github.com/Shopify/kubernetes-deploy/issues/51 (but I haven't looked into it)

File: README.md [hunk 14]
```diff
@@ -29,9 +29,13 @@ Requirements:
  - kubectl 1.5.1+ binary must be available in your path
  - `ENV['KUBECONFIG']` must point to a valid kubeconfig file that includes the context you want to deploy to
  - The target namespace must already exist in the target context
- - `ENV['GOOGLE_APPLICATION_CREDENTIALS']` must point to the credentials for an authenticated service account if your user's auth provider is gcp
+ - `ENV['GOOGLE_APPLICATION_CREDENTIALS']` must point to the credentials for an authenticated service account if your user's auth provider is GCP
  - `ENV['ENVIRONMENT']` must be set to use the default template path (`config/deploy/$ENVIRONMENT`) in the absence of the `--template-dir=DIR` option
 
+The tool also provides the task to restart all pods of a deployment by touching the ENV variable.
+The following command will fetch `web` and `jobs` deployments are make all pods to restart according to the rollout strategy:
```
Comment (KnVerey) [inline README.md]: Wording nitpicks:

> The tool also provides a task for restarting all of the pods in one or more deployments. It triggers the restart by touching the `RESTARTED_AT` environment variable in the deployment's podSpec. The rollout strategy defined for each deployment will be respected by the restart.

> The following command will restart all pods in the `web` and `jobs` deployments:

File: exe/kubernetes-deploy-restart (hunk 1 above)
Comment (KnVerey) [inline exe/kubernetes-deploy-restart]: I agree with this being a separate executable

File: exe/kubernetes-deploy-restart [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+require 'kubernetes-deploy/restart_task'
+
+raw_deployments = nil
+ARGV.options do |opts|
+  opts.on("--deployments=LIST") { |v| raw_deployments = v }
+  opts.parse!
+end
+
+if raw_deployments.nil?
+  puts "Failed: specify at least one deployment to restart with --deployments flag"
```
Comment (KnVerey) [inline exe/kubernetes-deploy-restart]: Very minor, but should we initialize the logger first and `logger.fatal` here?

Comment (KnVerey) [inline exe/kubernetes-deploy-restart]: And related to this, could this task use `KubernetesDeploy.logger` for output consistency with regular deploys, and sensitivity to `ENV["LEVEL"]` and `ENV["DEBUG"]`? It's in `lib/kubernetes-deploy.rb`, so it's actually already require'd.

File: lib/kubernetes-deploy/restart_task.rb (hunk 2 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/restart_task.rb]: Should this helper also move to `kubernetes-deploy/kubeclient_builder`? The tests could use it too, and some of the Kubernetes Resource classes will probably need the same when we convert them.

File: lib/kubernetes-deploy/restart_task.rb [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      @v1beta1_kubeclient.patch_deployment(
+        record.metadata.name,
+        build_patch_payload(record),
+        @namespace
+      )
+    end
+
+    def fetch_deployments(list)
+      list.map do |name|
+        records = @v1beta1_kubeclient.get_deployments(label_selector: "name=#{name}", namespace: @namespace)
```
Comment (KnVerey) [inline lib/kubernetes-deploy/restart_task.rb]: I think this code isn't handling the case where the deployment name is bad / not found for whatever reason. In any case I think we should use `@v1beta1_kubeclient.get_deployment(name, @namespace)` and possibly capture/re-raise the `KubeException` we'd get on 404 (if the resulting message would suck otherwise). Although Shopify's templates label all resources with a name label matching the resource name by convention, we don't actually control the templates, so we shouldn't rely on it (it's fine in tests since we do control those).

File: exe/kubernetes-deploy-restart [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  exit 1
+end
+
+logger = Logger.new(STDERR)
+logger.level = Logger::INFO
+logger.formatter = ->(severity, _datetime, _progname, msg) do
+  "[#{severity}] #{msg}\n"
+end
+
+restart = KubernetesDeploy::RestartTask.new(namespace: ARGV[0], context: ARGV[1], logger: logger)
```
Comment (KnVerey) [inline exe/kubernetes-deploy-restart]: Looks like if the context is invalid, we'll raise `"#{context} context must be configured in your KUBECONFIG (#{ENV['KUBECONFIG']}). "` in the client builder, but what does the failure look like when the namespace arg is either missing or invalid?

File: test/integration/restart_task_test.rb (hunk 12 above)
Comment (KnVerey) [inline test/integration/restart_task_test.rb]: Can we also get tests for:
- no name passed at all
- one good, one bad name passed

File: exe/kubernetes-deploy-restart (hunk 15 above)
Comment (kirs) [inline exe/kubernetes-deploy-restart]: 👍 I'll use KubernetesDeploy.logger

Regarding the `raw_deployments` check: IMO initialization errors are different from the runtime errors that are logged with `fatal` status

File: README.md (hunk 14 above)
Comment (kirs) [inline README.md]: Thanks!

**Comment by kirs** [comment]
Thanks for the review; all the point have been really helpful. I've updated the PR and addresses all the things, including the watch for rollout and extensive tests.

File: lib/kubernetes-deploy/restart_task.rb [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+      wait_for_rollout(deployments)
+      deployments.each do |record|
+        @logger.info "Successfully restarted #{record.metadata.name.inspect}"
+      end
+    end
+
+    private
+
+    def wait_for_rollout(deployments)
```
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: @KnVerey had to copypaste majority of the method to handle waiting for `Deployment` rollout

Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: But on the bright side I didn't need to modify the resource

Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: I'm a little uncomfortable with the copy-pasting here from the runner. There's some inconsistent vocabulary here in that for the deploy script, we call the class `Runner` but for restart we call it `RestartTask`. If we use the `Task` paradigm, we could rename `Runner` to `DeployTask` and then have a `Task` class they both inherit from that defines `wait_for_completion`.

File: lib/kubernetes-deploy/restart_task.rb [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+      @v1beta1_kubeclient = build_v1beta1_kubeclient(context)
+    end
+
+    def perform(deployments_names)
+      if deployments_names.empty?
+        raise ArgumentError, "#perform takes at least one deployment to restart"
+      end
+      deployments = fetch_deployments(deployments_names.uniq)
+      deployments.each do |record|
+        response = patch_deployment_with_restart(record)
```
Comment (sirupsen) [inline lib/kubernetes-deploy/restart_task.rb]: This is some of the first stuff that uses the API for mutation in this gem. Is that because there is a goal to move to the API completely long-term? Is there an API for, say, `kubectl apply`?

File: test/integration/restart_task_test.rb [hunk 20]
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+require 'test_helper'
+require 'kubernetes-deploy/restart_task'
+
+class RestartTaskTest < KubernetesDeploy::IntegrationTest
+  def test_restart
+    deploy_fixtures("basic", subset: ["configmap-data.yml", "web.yml.erb"])
+
+    deployment = fetch_deployment("web")
+    refute fetch_restarted_at(deployment), "no RESTARTED_AT env on fresh deployment"
```
Comment (sirupsen) [inline test/integration/restart_task_test.rb]: You have these two lines in a few places here `assert_fetched_at deployment("web")` might be convenient and a little more readable 👀

File: test/integration/restart_task_test.rb [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+
+    restart = KubernetesDeploy::RestartTask.new(
+      context: KubeclientHelper::MINIKUBE_CONTEXT,
+      namespace: @namespace,
+    )
```
Comment (sirupsen) [inline test/integration/restart_task_test.rb]: Perhaps you could instantiate this in `setup` as `@task`?

File: test/integration/restart_task_test.rb [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+    assert_raises(ArgumentError) do
+      restart.perform([])
+    end
+  end
+
+  def test_restart_not_existing_context
+    assert_raises(KubernetesDeploy::KubeclientBuilder::ContextMissignError) do
```
Comment (sirupsen) [inline test/integration/restart_task_test.rb]: There's a typo in `ContextMissgn` 😱

File: kubernetes-deploy.gemspec [hunk 23]
```diff
@@ -21,6 +21,7 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
   spec.add_dependency "activesupport", ">= 4.2"
+  spec.add_dependency "kubeclient", "~> 2.3"
```
Comment (sirupsen) [inline kubernetes-deploy.gemspec]: I suspect this version is different from what we used before. When I clone down this branch to attempt to run the tests I get:

```
macbookbro kubernetes-deploy (restart-task) $ rt test/integration/restart_task_test.rb 
Run options: --seed 51416

# Running:

/Users/sirup/src/github.com/Shopify/kubernetes-deploy/test/test_helper.rb:65:in `create_namespace': uninitialized constant Kubeclient::Namespace (NameError)
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/test/test_helper.rb:57:in `claim_namespace'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/test/test_helper.rb:44:in `run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:822:in `run_one_method'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:323:in `run_one_method'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:310:in `block (2 levels) in run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:309:in `each'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:309:in `block in run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:349:in `on_signal'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:336:in `with_info_handler'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:308:in `run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:158:in `block in __run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:158:in `map'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:158:in `__run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:135:in `run'
	from /Users/sirup/src/github.com/Shopify/kubernetes-deploy/vendor/bundle/gems/minitest-5.10.1/lib/minitest.rb:62:in `block in autorun'
```

Comment (sirupsen) [inline kubernetes-deploy.gemspec]: Could you try to blow away `vendor`, `bundle install`, and ensure it still works for you?

File: test/integration/restart_task_test.rb [hunk 24]
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+require 'test_helper'
+require 'kubernetes-deploy/restart_task'
+
+class RestartTaskTest < KubernetesDeploy::IntegrationTest
```
Comment (sirupsen) [inline test/integration/restart_task_test.rb]: Good test coverage! Suggestions for two more tests:

* `wait_for_rollout` fails to see the containers come back up (perhaps test this by making them sleep for too long?)
* Restarting after having already restarted, i.e. `RESTARTED_AT` is already set.

File: test/test_helper.rb [hunk 25]
```diff
@@ -28,6 +28,12 @@ def assert_logs_match(regexp)
       @logger_stream.rewind
       assert_match regexp, @logger_stream.read
     end
+
+    def assert_logs_match_once(regexp)
+      @logger_stream.rewind
+      count = @logger_stream.read.scan(regexp).count
+      assert_equal 1, count, "Expected #{regexp} to appear only once in the log, appear #{count} times"
+    end
```
Comment (sirupsen) [inline test/test_helper.rb]: Instead of introducing a new helper, I think you should consider adding a `times` argument to `assert_logs_match`. That's more similar to helpers we have elsewhere internally :)

File: test/integration/restart_task_test.rb [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+      )
+    end
+  end
+
+  def test_restart_not_existing_namespace
+    restart = KubernetesDeploy::RestartTask.new(
+      context: KubeclientHelper::MINIKUBE_CONTEXT,
+      namespace: "walrus"
+    )
+    assert_raises(KubernetesDeploy::RestartTask::DeploymentNotFoundError) do
```
Comment (KnVerey) [inline test/integration/restart_task_test.rb]: Hm, I think the distinction between the namespace not existing and the deployment not existing is very useful to the user. Can we validate the namespace before attempting to access the deployment so we can make that distinction?

File: lib/kubernetes-deploy/restart_task.rb (hunk 19 above)
Comment (kirs) [inline lib/kubernetes-deploy/restart_task.rb]: The way to restart with `kubectl` is:

```
export RESTART_DATE=$(date -u '+%Y-%m-%dT%k:%M:%S%z')
kubectl patch deployment web -n trashbin \
  -p '{"spec":{"template":{"spec":{"containers":[{"name":"app","env":[{"name":"RESTART","value":"'$RESTART_DATE'"}]}]}}}}'
```

The long-term plan is to move to the API, that's the reason why I didn't use `kubectl` here.

File: test/integration/restart_task_test.rb (hunk 21 above)
Comment (kirs) [inline test/integration/restart_task_test.rb]: I didn't want to put this to `setup` because the arguments to `RestartTask.new` are different across test cases.

File: test/test_helper.rb (hunk 25 above)
Comment (kirs) [inline test/test_helper.rb]: 👍

File: README.md [hunk 27]
```diff
@@ -29,9 +29,16 @@ Requirements:
...
  - The target namespace must already exist in the target context
- - `ENV['GOOGLE_APPLICATION_CREDENTIALS']` must point to the credentials for an authenticated service account if your user's auth provider is gcp
+ - `ENV['GOOGLE_APPLICATION_CREDENTIALS']` must point to the credentials for an authenticated service account if your user's auth provider is GCP
  - `ENV['ENVIRONMENT']` must be set to use the default template path (`config/deploy/$ENVIRONMENT`) in the absence of the `--template-dir=DIR` option
 
+The tool also provides a task for restarting all of the pods in one or more deployments.
+It triggers the restart by touching the `RESTARTED_AT` environment variable in the deployment's podSpec.
+The rollout strategy defined for each deployment will be respected by the restart.
+
+The following command will restart all pods in the `web` and `jobs` deployments:<Paste>
```
Comment (kirs) [inline README.md]: My vim thing. Thanks for spotting it!

File: test/integration/restart_task_test.rb (hunk 26 above)
Comment (kirs) [inline test/integration/restart_task_test.rb]: Sure 👍

**Comment by kirs** [comment]
It's ready for another review.
The CI is green: https://buildkite.com/shopify/kubernetes-deploy-gem/builds/100#ce42d94a-e1cd-4bc3-8f50-e9308f10750d

File: README.md [hunk 28]
```diff
@@ -29,9 +29,16 @@ Requirements:
...
+ - `ENV['GOOGLE_APPLICATION_CREDENTIALS']` must point to the credentials for an authenticated service account if your user's auth provider is GCP
  - `ENV['ENVIRONMENT']` must be set to use the default template path (`config/deploy/$ENVIRONMENT`) in the absence of the `--template-dir=DIR` option
 
+The tool also provides a task for restarting all of the pods in one or more deployments.
+It triggers the restart by touching the `RESTARTED_AT` environment variable in the deployment's podSpec.
+The rollout strategy defined for each deployment will be respected by the restart.
+
+The following command will restart all pods in the `web` and `jobs` deployments:
+
+`kubernetes-deploy-restart <kube namespace> <kube context> --deployments=web,jobs`
```
Comment (KnVerey) [inline README.md]: Naming nit: I noticed @xldenis's WIP names that task  `kubernetes-run`... wdyt about this being `kubernetes-restart`?

File: kubernetes-deploy.gemspec (hunk 23 above)
Comment (KnVerey) [inline kubernetes-deploy.gemspec]: fwiw I just pulled the latest version of this branch and was able to run the tests fine.

**Comment by kirs** [comment]
Great suggestion. `kubernetes-restart` makes more sense 👍

**Comment by kirs** [comment]
This should unblock @xldenis to continue his work.
